larg4/Core/larg4Main_module.cc

```diff
@@ -148,70 +150,67 @@ larg4::larg4Main::larg4Main(fhicl::ParameterSet const & p)
+  if (visSpecificEvents_) {
+    std::vector<int> eventsToDisplayVec =
+            p.get<vector<int>>("eventsToDisplay");
+    for (size_t i = 0; i < eventsToDisplayVec.size(); i++) {
+      eventsToDisplay_[eventsToDisplayVec[i]] = true;
+    }
+    // Would be nice to have error checking here, but for now, if you
+    // do something silly, it'll probably just crash.
+  }
 
   // We need all of the services to run @produces@ on the data they will store. We do this
   // by retrieving the holder services.
   art::ServiceHandle<ActionHolderService> actionHolder;
   art::ServiceHandle<DetectorHolderService> detectorHolder;
 
-    detectorHolder->initialize();
-    // Build the detectors' logical volumes
-    detectorHolder -> constructAllLVs();
-    // And running @callArtProduces@ on each
-    actionHolder -> callArtProduces(producesCollector());
-    detectorHolder -> callArtProduces(producesCollector());
-
-    //    ((artg4tk::SteppingActionBase*)&*pla)-> callArtProduces(this);
-    // ((artg4tk::EventActionBase*)&*pla) -> callArtProduces(this);
-    // ((artg4tk::TrackingActionBase*)&*pla) -> callArtProduces(this);
-  // Set up the random number engine.
-  // See the documentation in RandomNumberHeader.h for
-  // how this works. Note that @createEngine@ is a member function
-  // of our base class (actually, a couple of base classes deep!).
-  // Note that the name @G4Engine@ is special.
-  if (seed_ == -1) {
-	  // Construct seed from time and pid. (default behavior if
-	  // no seed is provided by the fcl file)
-	  // Note: According to Kevin Lynch, the two lines below are not portable.
-	  seed_ = time(0) + getpid();
-	  seed_ = ((seed_ & 0xFFFF0000) >> 16) | ((seed_ & 0x0000FFFF) << 16); //exchange upper and lower word
-	  seed_ = seed_ % 900000000; // ensure the seed is in the correct range for createEngine
+  detectorHolder->initialize();
+  // Build the detectors' logical volumes
+  detectorHolder -> constructAllLVs();
+  // And running @callArtProduces@ on each
+  actionHolder -> callArtProduces(producesCollector());
+  detectorHolder -> callArtProduces(producesCollector());
+
+  // ((artg4tk::SteppingActionBase*)&*pla)-> callArtProduces(this);
+  // ((artg4tk::EventActionBase*)&*pla) -> callArtProduces(this);
+  // ((artg4tk::TrackingActionBase*)&*pla) -> callArtProduces(this);
+
+  // -- Check for invalid seed value
+  if (seed_ > 900000000) {
+    //mf::LogError("SeedCheck") << "Bad seed provided, max seed value is 9E8.
+    throw cet::exception("largeant:BadSeedValue")
+          << "The provided largeant seed value: " << seed_
+          << " is invalid! Maximum seed value is 9E8.";
   }
-  createEngine( seed_, "G4Engine");
+  // Set up the random number engine.
+  // -- D.R.: Use the NuRandomService engine for additional control over the seed generation policy
+  (void)art::ServiceHandle<rndm::NuRandomService>()->createEngine(*this,"G4Engine",p,"seed");
```

Comment:
Is there a reason you cannot simply call:
```.cpp
createEngine(p.get<long>("seed", 0), "G4Engine");
```
or some other default seed value.  If that function call is sufficient, then that's a lot simpler than introducing a dependency on `NuRandomService`, which has known multi-threading problems.

Reply:
The reason to use the NuRandomService `createEngine` was to have control over the engine policy. We are able to reproduce events in LegacyLArG4 following this [redmine wiki](https://cdcvs.fnal.gov/redmine/projects/larsoft/wiki/Reproducing_jobs_using_the_same_random_number_sequences) with the `perEvent` policy.

We could use the native `createEngine` from the EDProducer base class; however, there are multiple wikis that say things like

> The preferred way to control random number streams is through NuRandomService

I wasn't aware that there were issues with multi-threading for the NuRandomService, however, and I know that @hanswenzel (and maybe others) have started looking into multi-threading.

Reply:
Thanks for the info, @drivera-fnal.  We should re-evaluate the guidance given in the redmine documentation...from a multi-threading perspective `NuRandom` is *not* the preferred way of handling random numbers.  

Let me talk with the SciSoft team regarding the best way forward here.  Adding `NuRandom` right now is suspect.  Do you mind opening up a support issue, where you mention what you require in terms of random numbers/reproducibility?

https://cdcvs.fnal.gov/redmine/projects/larsoft/issues/new

Thanks.

Reply:
No problem @knoepfel 
I will fill out a new support issue with the details of our request.